src/main/java/htsjdk/tribble/index/IndexFactory.java

```diff
@@ -73,47 +68,41 @@
      * an enum that contains all of the information about the index types, and how to create them
      */
     public enum IndexType {
-        LINEAR(LinearIndex.MAGIC_NUMBER, LinearIndex.INDEX_TYPE, LinearIndexCreator.class, LinearIndex.class, LinearIndexCreator.DEFAULT_BIN_WIDTH),
-        INTERVAL_TREE(IntervalTreeIndex.MAGIC_NUMBER, IntervalTreeIndex.INDEX_TYPE, IntervalIndexCreator.class, IntervalTreeIndex.class, IntervalIndexCreator.DEFAULT_FEATURE_COUNT),
+        LINEAR(AbstractIndex.MAGIC_NUMBER, LinearIndex.INDEX_TYPE, true, LinearIndex::new, LinearIndexCreator.DEFAULT_BIN_WIDTH),
+        INTERVAL_TREE(AbstractIndex.MAGIC_NUMBER, IntervalTreeIndex.INDEX_TYPE, true, IntervalTreeIndex::new, IntervalIndexCreator.DEFAULT_FEATURE_COUNT),
         // Tabix index initialization requires additional information, so generic construction won't work, thus indexCreatorClass is null.
-        TABIX(TabixIndex.MAGIC_NUMBER, null, null, TabixIndex.class, -1);
+        TABIX(TabixIndex.MAGIC_NUMBER, null, false, TabixIndex::new, -1) ;
 
         private final int magicNumber;
         private final Integer tribbleIndexType;
-        private final Class<IndexCreator> indexCreatorClass;
         private final int defaultBinSize;
-        private final Class<Index> indexType;
+        private final boolean canCreate;
+        private final IOFunction<InputStream, Index> createFromInputStream;
 
         public int getDefaultBinSize() {
             return defaultBinSize;
         }
 
-        public IndexCreator getIndexCreator() {
-            try {
-                return indexCreatorClass.newInstance();
-            } catch ( final InstantiationException | IllegalAccessException e ) {
-                throw new TribbleException("Couldn't make index creator in " + this, e);
-            }
-        }
-
         public boolean canCreate() {
-            return indexCreatorClass != null;
+            return canCreate;
         }
 
-        IndexType(final int magicNumber, final Integer tribbleIndexType, final Class creator, final Class indexClass, final int defaultBinSize) {
+        private interface IOFunction<T,R>{ R apply(T t) throws IOException;}
```

Comment:
This would be easier to read if it's formatted across lines
```
private interface IOFunction<T, R> {
    R apply(T t) throws IOException;
}
```

Also isn't `T` `InputStream` and `R` is always a subtype of `AbstractIndex`? Is there a reason to use a type abstraction here?

Reply:
If I call it IOFunction and make it abstract then I don't have to think of an name for the concretized type.  I'll change.